Getting started guide for testing (#1394)
o3de/o3de.org: content/docs/user-guide/testing/getting-started/_index.md

---
linkTitle: Getting Started
title: Getting Started with Test Automation
description: Learn about the automated testing tools provided with Open 3D Engine (O3DE).
toc: true
weight: 500
---

This guide introduces the automated testing tools in **Open 3D Engine (O3DE)** and covers their basic usage. More detailed information on each tool exists on their specific documentation pages.

## Overview

> O3DE uses [CMake](https://cmake.org/cmake/help/latest/), a build system that includes [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html). CTest is a generic test runner tool which coordinates executing and reporting on tests. O3DE uses CTest to start all automated tests of the engine and tools. New projects using O3DE can also register and run their own tests through CTest, to test their unique game code or engine extensions. CTest can be executed on a developer's local machine to verify code health. It also gets executed in the O3DE Automated Review (AR) pipeline to help prevent bad merges. The sections below introduce the primary way of starting CTest, and describes how to register tests.

Author: Used 'committing changes' instead of 'any submission'

![CMake and CTest workflow](/images/user-guide/testing/getting-started/cmake_ctest_workflow.png)

## CTest

[CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html) is similar to other test frameworks like GoogleTest, PyTest, or JUnit; though it uses a higher-level generic interface. CTest registers each test as a string of commandline arguments, and reports on whether invoking those arguments succeeded (returned zero) without hanging. CTest can also coordinate running tests in parallel processes. Unlike other test runners, CTest is programming language agnostic by interfacing with the OS shell. However this means a notable feature is absent from CTest: other test frameworks typically provide a lower-level interface which directly invokes functions inside a specific programming language. To provide this feature, CTest is used to invoke lower-level test runners.

> To simplify using function-level test reporting, O3DE provides wrapper code to register tests for frameworks such as GoogleTest and PyTest into CTest. These wrappers enable writing tests which run on any operating system. XML files are generated to track the low-level results not reported in CTest, as well as artifacts such as log output and crash dumps.

Author: Comma not added, instead updated to "tests which can run on any operating system"

> To simplify using function-level test reporting, O3DE provides wrapper code to register tests for frameworks such as GoogleTest and PyTest into CTest. These wrappers enable writing tests which run on any operating system. XML files are generated to track the low-level results not reported in CTest, as well as artifacts such as log output and crash dumps.

Reviewer: In this context, where "can run on any operating system" describes "tests", then I believe "which" should follow a comma. Alternatively, you can use "that" without a comma. So the following two sentence structures are correct: 
- These wrappers enable writing tests, which can run on any operating system.
- These wrappers enable writing tests that can run on any operating system.

> In short, all tests in O3DE use CTest as a high level test coordinator. When executed, CTest calls other lower-level test tools and then reports on their success.

Reviewer: ```suggestion
In short, all tests in O3DE use CTest as a high-level test coordinator. When executed, CTest calls other lower-level test tools and then reports on their success.
```

![CTest calls other test runners](/images/user-guide/testing/getting-started/ctest_to_runners.png)

### Starting CTest

CTest expects its working directory to be a CMake build directory, so be sure to first navigate to this directory:

```shell
cd <local_path_to>/o3de/build/<build_folder>
ctest -C <build_configuration>
```

> You may not want to run the commands above without also filtering the set of tests to run. Without a filter the command will run every test, and likely result in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing `CTRL + C` . The most commonly used suites are Smoke and Main, which have requirements to execute fast. The following ctest command filters to only tests in these two suites:

Author: This is, in part, why I mention cancelling the run. I'm happy to change the order if it's an issue.

> You may not want to run the commands above without also filtering the set of tests to run. Without a filter the command will run every test, and likely result in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing `CTRL + C` . The most commonly used suites are Smoke and Main, which have requirements to execute fast. The following ctest command filters to only tests in these two suites:
> 
> ```shell
> cd <local_path_to>/o3de/build/<build_folder>
> ctest -C <build_configuration> -L "(SUITE_smoke|SUITE_main)"
> ```
> 
> The build configuration should be one you have already built locally. Running the Smoke and Main suites for the `profile` build configuration on a Linux machine should look like:
> 
> ```shell
> cd user/github/o3de/build/linux
> ctest -C profile -L "(SUITE_smoke|SUITE_main)"
> ```

Reviewer 2: Suggestion to break up and reformat slightly in a way that might resolve some of this confusion and reduce redundancy. It's not perfect but it feels like a better flow to me:

````suggestion
CTest can also take a set of tests to run with the `-L` argument. These test suites should be contained in parenthesis (`(..)`) and have their names separated with a `|` character. The following example demonstrates this syntax for running the Main and Smoke suites for a `profile` build:

```shell
cd user/github/o3de/build/linux
ctest -C profile -L "(SUITE_smoke|SUITE_main)"
```

The Smoke and Main suites are required to be quick. Unless performing a full regression, performance, or project-specific test, we recommend running these suites.

{{< caution >}}
Without a filter the command will run every test, and likely result in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing `CTRL+C`.
{{< /caution >}}
````

Author: Good suggestion. Moved the caution to be at the top of this block.

> CTest expects its working directory to be a CMake build directory, so be sure to first navigate to this directory:
> 
> ```shell
> cd <local_path_to>/o3de/build/<build_folder>
> ctest -C <build_configuration>
> ```
> 
> You may not want to run the commands above without also filtering the set of tests to run. Without a filter the command will run every test, and likely result in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing `CTRL + C` . The most commonly used suites are Smoke and Main, which have requirements to execute fast. The following ctest command filters to only tests in these two suites:
> 
> ```shell
> cd <local_path_to>/o3de/build/<build_folder>
> ctest -C <build_configuration> -L "(SUITE_smoke|SUITE_main)"
> ```
> 
> The build configuration should be one you have already built locally. Running the Smoke and Main suites for the `profile` build configuration on a Linux machine should look like:
> 
> ```shell
> cd user/github/o3de/build/linux
> ctest -C profile -L "(SUITE_smoke|SUITE_main)"
> ```
> 
> After running CTest, results save to `.../<build_folder>/Testing/`.  If you prefer to see a summary of failures in your terminal, add the flag `--output-on-failure`.

Reviewer: I have some suggestions for organizing this section. 
- Split up the instructions for Windows and Linux. This makes it easier for users to find the instructions for their machine. Also, by using **On Windows** and **On Linux** headers, you can cross link to them if you need to in future docs. 
- Write the instructions in steps (1, 2, etc.)
   - Created Step 2: "Use CTest to run your tests." Then, put the two instructions for running CTest with filters and without into sub-bullets. 
- Moved the statement about `<build_configuration>` after you first introduce it. 

Visually, the suggestion would look like this: 
![image](https://user-images.githubusercontent.com/75444793/153935365-76315e77-3bac-4d92-9a61-4fd495b4bf92.png)

```suggestion

#### On Windows

1. In a terminal, navigate to the CMake build directory, which is CTest's expected working directory.

    \```shell
    cd <local_path_to>/o3de/build/<build_folder>
    \```

2. Use CTest to run your tests. 
   
   - You can run the following `ctest` command. This runs every test and likely results in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing **CTRL**+**C**.

        \```shell
        ctest -C <build_configuration>
        \```

        - `<build_configuration>` should be one that you've already built locally.

    - Alternatively, you can filter the set of tests to run. The most commonly used suites are Smoke and Main, which have requirements to execute fast. The following `ctest` command filters to only test in these two suites:

        \```shell
        cd <local_path_to>/o3de/build/<build_folder>
        ctest -C <build_configuration> -L "(SUITE_smoke|SUITE_main)"
        \```

#### On Linux

For a Linux machine, use the following commands to navigate to the `linux` CMake build directory and run the Smoke and Main suites for the `profile` build configuration.

\```shell
cd user/github/o3de/build/linux
ctest -C profile -L "(SUITE_smoke|SUITE_main)"
\```

After running CTest, the results save to `.../<build_folder>/Testing/`.  If you prefer to see a summary of failures in your terminal, run the `ctest` command with the the flag `--output-on-failure`.
```

Reviewer 3: These are some great suggestions.

Author: This is an interesting suggestion, though there are far more similarities on Windows and Linux than differences. Trying to find a narrower split to describe only the differences.

Reviewer: Another way is for each step, specify **Windows** or **Linux** if they are different, or don't specify either if they're the same. 
For example: 

1. In a terminal, navigate to the CMake build directory, which is CTest's expected working directory.
   - **Windows**:
      ```shell
      cd <local_path_to>/o3de/build/<build_folder>
      ```
   - **Linux**:
      ```shell
      cd user/github/o3de/build/linux
      ```
2. Use CTest to run your tests. The commands are similar for both Windows and Linux.
   - You can run the following `ctest` command. This runs every test and likely results in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing **CTRL**+**C**.
        ```shell
        ctest -C <build_configuration>
        ```
        - `<build_configuration>` should be one that you've already built locally.
    - Alternatively, you can filter the set of tests to run. The most commonly used suites are Smoke and Main, which have requirements to execute fast. The following `ctest` command filters to only test in these two suites:
        ```shell
        ctest -C <build_configuration> -L "(SUITE_smoke|SUITE_main)"
        ```

For more information on CTest usage, refer to its [online documentation](https://cmake.org/cmake/help/latest/manual/ctest.1.html).

### Adding test modules to CTest

> CTest registers entire modules of test code. These modules typically contain multiple individual tests of the same feature. Adding new tests has three major steps, with a prerequisite:

Reviewer 4: It may be worth pointing out that if you're simply writing new tests, most of these steps do not need to be done.

Author: "Writing new tests" is a pretty vague motivation, though I think this statement would apply to "expanding existing tests"

The following suggestions do suggest where to create/look-for this file, such as Line 127, so I didn't think it necessary to add here at the top.

0. Add a build target for the production code which tests will target.
1. Add a build target for tests which need to be compiled (not required for Python).
2. Register your test module in a `CMakeLists.txt`.
3. Add individual test functions into the test module.

Specific steps vary for different types of tests, discussed in later sections.

## GoogleTest

> Many tests are easiest to write in the same language as the production code they target. Since the majority of O3DE uses C++, the majority of its tests also use C++. O3DE's C++ tests are typically small [unit level](https://softwaretestingfundamentals.com/unit-testing/) tests of specific low-level functionality, which run extremely fast. O3DE uses [GoogleTest](https://github.com/google/googletest/blob/main/docs/index.md) to coordinate these tests, with an execution wrapper named AzTestRunner. CTest will call AzTestRunner to load the test library and the target library, and then execute any loaded GoogleTest tests.

Reviewer: Suggestions to make sentences flow better and fix style syntax.
```suggestion
For many tests, it's easier to write them in the same language as the production code that they target. Since the majority of O3DE uses C++, the majority of its tests also use C++. O3DE's C++ tests are typically small [unit level](https://softwaretestingfundamentals.com/unit-testing/) tests of specific low-level functionality, which run extremely fast. O3DE uses [GoogleTest](https://github.com/google/googletest/blob/main/docs/index.md) to coordinate these tests, with an execution wrapper named `AzTestRunner`. CTest calls `AzTestRunner` to load the test library and the target library, and then execute any loaded GoogleTest tests.
```

> When building production C++ code for O3DE features, that code gets built into a library to later load into the engine. When writing C++ tests for this code, the test-code should be built into a separate library with a dependency on the production library. Building separate test libraries has the advantage of keeping production binaries lean and easy to ship, free of extraneous ```#if defined``` blocks for test code, and ensures we test the exact same interfaces that we ship.

Reviewer: Suggestions to simplify sentences and make them flow better.
```suggestion
Production C++ code for O3DE features gets built into a library, which can later be loaded into the engine. When writing C++ tests for this code, the test-code should be built into a separate library with a dependency on the production library. Building separate test libraries has the advantage of keeping production binaries lean and easy to ship and free of extraneous ```#if defined``` blocks for test code. It also ensures we test the exact same interfaces that we ship.
```

Reviewer 2: Incorporating Chanelle's suggestions with some other changes:

```suggestion
Production C++ code for O3DE features gets built into a library, which can later be loaded into the engine. When writing C++ tests for this code, the test-code should be built into a separate library with a dependency on the production library. Building separate test libraries has the advantage of keeping production binaries lean and easy to ship and free of extraneous ```#if defined``` blocks for test code. It also ensures test interfaces are identical to production interfaces.
```

The use of 2nd person in documentation ("we") should generally be avoided for a number of reasons, but primarily because the usage of this term implies the endorsement of the O3DF or O3DE project owners. This is why it's used only in _best practices_-style writing ("we recommend" i.e. "the O3DE project recommends") regarding neutral technical requirements that fit with efficient project usage.

Author: Accepted, though revised as interfaces for test libraries and production libraries are not identical.

ly_add_target(
    NAME MyModuleName
    NAMESPACE Gem
    FILES_CMAKE
        mymodule_files.cmake
    INCLUDE_DIRECTORIES
        PRIVATE
            Source
        PUBLIC
            Include
    BUILD_DEPENDENCIES
        PRIVATE
            Legacy::CryCommon
    COMPONENT
        MyModuleComponent
)

# New test module registration from Step 1
# this configures CMake to also build a library called MyModuleName.Tests.so (.dll on Windows)
# the rest of CMake will refer to this freshly created module by its NAME ("MyModuleName.Tests")
if(PAL_TRAIT_BUILD_SUPPORTS_TESTS)  # only create these test modules if the target platform supports testing

    ly_add_target(
        NAME MyModuleName.Tests SHARED
        NAMESPACE Gem
        FILES_CMAKE
            mymodule_test_files.cmake
        INCLUDE_DIRECTORIES
            PRIVATE
                Tests
                Source
        BUILD_DEPENDENCIES
            PRIVATE
                AZ::AzTest
                Legacy::CryCommon
                Gem::MyModuleName
        COMPONENT
            MyModuleComponent
    )

    # Register new test module from Step 2
    ly_add_googletest(
        NAME MyModuleName.Tests
    )

endif()
```

> #### Step 0: Add a production build target

Reviewer: Similar to earlier suggestion, change step 0 to "Prerequisite". 
```suggestion
#### Prerequisite: Add a production build target
```

> Before configuring tests, you should first define the library which gets tested. This is defined in a `CMakeLists.txt`, which typically lives in the same directory as the code or in a parent directory. For more information on configuring `CMakeLists.txt`, please read more about adding library targets at [Getting Started with the CMake Build System](/docs/user-guide/build/). Note that you only need to define a library that will be built; you can start configuring and writing tests before you finish writing the production code.

Reviewer: ```suggestion
Before configuring tests, you must first define the library that you want to test. This is defined in a `CMakeLists.txt`, which lives in the same directory as the code or in a parent directory. For more information on configuring `CMakeLists.txt`, please read more about adding library targets at [Getting Started with the CMake Build System](/docs/user-guide/build/). Note that you only need to define the library that you want to build. You can start configuring and writing tests before you finish writing the production code.
```
- Use "must" for required actions, and "should" for recommendations or probabilities. 
- Changed to "that you want to test" and "that you want to build." to clarify that the developer is the one choosing which library to test and build.

Reviewer 2: Regarding "should"/"must" usage - we try and write to [IETF 2119](https://datatracker.ietf.org/doc/html/rfc2119) as this is what programmers interpret these words as (consciously or not - they've been reading them with these meanings their whole career.)

Author: The ordering a grey-area to me, but I think the recommendation is strong enough to default to a must.

#### Step 1: Add a test build target

> Similar to the production build target, the test target defines a library in a `CMakeLists.txt` configuration file. Start by finding the `CMakeLists.txt` for your module, which should already be created from Step 0. It should exist at a path similar to `o3de/.../<MyModule>/CMakeLists.txt`.

Reviewer: ```suggestion
Similar to the production build target, the test target defines a library in a `CMakeLists.txt` configuration file. Start by finding the `CMakeLists.txt` for your module, which you created in the prerequisite step. It should exist at a path similar to `o3de/.../<MyModule>/CMakeLists.txt`.
```

> Modify the CMakeLists.txt file to define your new test module, using the helper function `ly_add_target()`. Similar to the production build target, it is easiest to create another file which lists the C++ files used to compile the test library. The example above uses `dev/.../<MyModule>/mymodule_test_files.cmake`, which would have content similar to:

Reviewer: ```suggestion
Modify the `CMakeLists.txt` file to define your new test module by using the helper function `ly_add_target()`. Similar to the production build target, it's easiest to create another file that lists the C++ files used to compile the test library. The example above uses `dev/.../<MyModule>/mymodule_test_files.cmake`, which has content similar to the following:
```

```
set(FILES
    test/MyModuleTestFile.cpp
    test/MyModuleMathTests.cpp
)
```

#### Step 2: Register the test module

> The final `CMakeLists.txt` step is to register the module with CTest. For this, use helper function `ly_add_googletest()`

Reviewer: ```suggestion
In `CMakeLists.txt`, register the module with CTest by using the helper function `ly_add_googletest()`.
```
You can change "final `CMakeLists.txt` step" to "In `CMakeLists.txt,` ..." which allows you to restructure the rest of the sentence so its more concise. Also, i think it's clearer to describe the step as taking place "In `CMakeLists.txt`".

> To verify everything is set up correctly, run the [CMake configure command](/docs/user-guide/build/configure-and-build/) from **CMake CLI** or **CMake GUI**. This will register everything you just added, and emit errors if anything was misconfigured.

Reviewer: ```suggestion
To verify everything is set up correctly, run the [CMake configure command](/docs/user-guide/build/configure-and-build/) from **CMake CLI** or **CMake GUI**. This registers everything you just added and emits errors, if anything was misconfigured.
```

    EXPECT_TRUE(someResult);
    EXPECT_FALSE(secondaryProperty);
}
```

For more information on writing tests in C++, see [Using AzTest](/docs/user-guide/testing/aztest/aztest/).

## GoogleBenchmark

For performance benchmarks of small pieces of C++ code O3DE uses [GoogleBenchmark](https://github.com/google/benchmark/blob/main/docs/index.md), which is quite similar to GoogleTest. However, the definition of failure is a major difference between a normal test and a benchmark. In most tests a pass/fail status is directly evaluated to a boolean state, creating an objective report of success and failure. Benchmarks instead create a subjective performance metric. These metrics are most valuable when periodically recorded, to help detect trends over time. The only objective failure during a benchmark occurs when the code fails to run or crashes.

The steps for configuring a GoogleBenchmark library are identical to the [steps above for GoogleTest libraries](#googletest), with a couple exceptions. The include statement is `#include <benchmark/benchmark.h>` and the cmake helper function is:

```
ly_add_googlebenchmark(
    NAME MyModule.Benchmarks
)
```

## PyTest

> Certain tests are easier to write in a scripting language like Python. The scope of these tests are typically at the [integration level](https://softwaretestingfundamentals.com/integration-testing/), verifying mid-to-high level functionality of workflows. This type of test helps verify system correctness similar to how an end-user will experience the software. Despite that positive aspect, these tests are typically slow and provide less specific failure information. Not all Python-based tests are integration tests: any Python-based tools will prefer unit tests written in Python. For all tests written in Python, O3DE uses [PyTest](https://docs.pytest.org/).

Reviewer: ```suggestion
Some tests are easier to write in Python, such as unit tests for Python-based tools or [integration tests](https://softwaretestingfundamentals.com/integration-testing/). For tests written in Python, O3DE uses [PyTest](https://docs.pytest.org/). Integration tests verify mid-to-high level functionality of workflows. It verifies the system correctness, similar to how an end-use experiences the software. Despite that, integration tests are typically slow, and they provide less specific failure information.
```

- Remove "in a scripting language like..." if we are only talking about tests written in Python. If the developer can write tests with other scripting languages and use PyTest with them, you can clarify that in another sentence. 
- Favor "some" over "certain", because certain as a determiner means having been determined but unspecified. When in this paragraph, you do specify at least some of the types of tests.
- Moved the last two sentences to the beginning of the paragraph since they are more general and introductory-level statements. 
- **Can you verify** if it's accurate to call the tests "integration test"? Specifying "integration test" throughout the paragraph helps clarify what tests you're talking about. Otherwise you could be talking about integration tests or tests written in Python.

Author: The first sentence is attempting to introduce a general testing concept, and then clarify that Python + PyTest is used in O3DE. Debating on whether that is an important mental-bridge to cross, or necessarily gets in the way as you describe.

Will clarify language around test-types.

> While interpreted languages like Python can have low performance, these tests should not perform heavy computation in Python. Tests should instead focus on coordinating workflows: signalling events and then verifying a response. Leave heavy operations to the code targeted by a test, and perform simple checks in Python test-code.

Reviewer: ```suggestion
While interpreted languages like Python can have low performance, tests written in Python should not perform heavy computation. Instead, the tests should focus on coordinating workflows: for example, signalling events and then verifying a response. Leave heavy operations in the code that's targeted by the test, and perform simple checks in the Python test-code.
```

There are often two Python interpreter instances used during O3DE tests. This can complicate determining which interpreter executes your code, so pay close attention to how the test is defined:

- Test exists outside of **O3DE Editor**
  - Use in generic tests that launch applications and send external signals.
  - Helps monitor for application crashes.
  - Test file names should start with `test_`.
  - Uses the same Python interpreter that launches when you run `o3de/python/python.sh` or `o3de\python\python.cmd`.

- Test embedded inside O3DE Editor
  - Use in `EditorPythonBindings` tests which target workflows inside O3DE Editor.
  - Helps expose editor functionality to a test script.
  - Test file names should ***not*** include affixes such as `test_` or `_tests`.
  - PyTest is not used inside the editor, and most PyTest functionality is not available

>     - PyTest is still initiated by the external Python interpreter instance, to handle Editor crashes.

Reviewer: I'm a bit confused by this part. Overall the first paragraph seems to be about the two Python interpreter instances, though I'm not sure which they are. I also understand that I need to see how the test is defined to determine which interpreter executes my code. So in the following list of points, I expect to learn about the two Python interpreter instances, and the different definitions of tests and what kind of interpreter they use. 

Instead I see that there are two ways a test is defined: outside of O3DE Editor and inside of O3DE Editor, and a list of characteristics of those two types of tests. However, I did learn that tests outside of O3DE Editor "use the same Python interpreter that launches when you run `o3de/python/python.sh` or `o3de\python\python.cmd.`" But still I'm not sure if the list is fitting for the first paragraph introducing it.

Author: I am also unsure how to clarify this further, but am open to suggestions. Perhaps a prescriptive list works better than a descriptive list? For now, I'll add links to existing documentation on LyTestTools and EditorTest, though I don't think this fully resolves the complex description.

Worst case, this confusion means the reader is paying attention to the complexity.

> There are often two Python interpreter instances used during O3DE tests. This can complicate determining which interpreter executes your code, so pay close attention to how the test is defined:
> 
> - Test exists outside of **O3DE Editor**
>   - Use in generic tests that launch applications and send external signals.
>   - Helps monitor for application crashes.
>   - Test file names should start with `test_`.
>   - Uses the same Python interpreter that launches when you run `o3de/python/python.sh` or `o3de\python\python.cmd`.
> 
> - Test embedded inside O3DE Editor
>   - Use in `EditorPythonBindings` tests which target workflows inside O3DE Editor.
>   - Helps expose editor functionality to a test script.
>   - Test file names should ***not*** include affixes such as `test_` or `_tests`.
>   - PyTest is not used inside the editor, and most PyTest functionality is not available
>     - PyTest is still initiated by the external Python interpreter instance, to handle Editor crashes.

Reviewer 2: My attempt:

```suggestion
O3DE can run multiple Python interpreter instances for testing. Currently, there are two interpreters: The _Editor interpreter_ and the _external interpreter_. It can be difficult to determine which environment your Python tests execute in. To determine which environment your test should run in, and how to execute it there, follow the remainder of this section.

#### Tests in the external interpreter

The external interpreter runs outside of any active O3DE application. This is the same interpreter that launches when you run `o3de/python/python.sh` or `o3de\python\python.cmd`., and is best used for tests which involve the following:

- Generic tests that launch applications and send external signals.
- Monitoring application crashes.

To integrate with the external interpreter, your tests must be in filenames which start with `test_`.

#### Tests in the Editor interpreter

The Editor interpreter runs inside of O3DE Editor, using an internal Python library. Although this environment isn't equivalent to launching `o3de/python/python.*`, it uses the same version of the Python interpreter. The Editor interpreter is best used for tests which involve the following:

- Targeting specific functionality within the Editor or another application
- Exposing Editor functionality to another test

To integrate with the Editor interpreter, create a test which uses `EditorPythonBindings`. These tests **must not** be in a file starting with `test_` or `tests_`.

{{< note >}}
PyTest is not used within the Editor, and as a consequence PyTest functionality is unavailable to tests which run in the Editor interpreter. If your test requires the usage of PytTest, create a test which runs in the external interpreter and then monitors activity appropriately as reported by a test which runs in the Editor interpreter.
{{< /note >}}
```

The language used here for some of the details of in-Editor interpreter testing are what I was able to divine from the content as written; if they're not precise, make sure to update.

Author: Good suggestions, updating for accuracy

### Registering a new Python test

> Registering a Python-based test is simpler than registering a C++ test. However it still requires defining a C++ library that will be tested, which is often already complete before designing integration tests. The steps below assume you have already defined a production library. You can read more about defining production code in [Getting Started with the CMake Build System](/docs/user-guide/build/).

Reviewer: ```suggestion
Registering a Python-based test is simpler than registering a C++ test. However, it still requires that you define the C++ library that you want to test, which you may have already completed before designing integration tests. The steps below assume you have already defined a production library. You can read more about defining production code in [Getting Started with the CMake Build System](/docs/user-guide/build/).
```

#### Step 1: Register a PyTest target

> Find the `CMakeLists.txt` that defines the system you are testing. It should exist at a path similar to `o3de/.../<MyModule>/CMakeLists.txt`. Note that it may be appropriate to register broader tests in a parent directory's CMakeLists.txt instead. If you are creating a test that is specific to a game project, you should register it in a `CMakeLists.txt` in that game project. When you find the right place for the test to be registered, add a line similar to:

Author: I disagree that combining the first two sentences increases legibility. Changed the content after the colon to be a bulleted list of reasons.

```
ly_add_pytest(
    NAME myPythonTest 
    PATH ${CMAKE_CURRENT_LIST_DIR}/Tests/test_MyFile.py
    TEST_SERIAL # many larger-than-unit tests can interfere with one another in parallel
    RUNTIME_DEPENDENCIES # helps the test build-and-run from an IDE
        Legacy::Editor
        AZ::AssetProcessor
        AutomatedTesting.Assets
)
```

> To verify everything is set up correctly, run the [CMake configure command](/docs/user-guide/build/configure-and-build/) from the CMake CLI or CMake GUI. This will register everything you just added, and emit errors if anything was misconfigured.

Reviewer: ```suggestion
To verify everything is set up correctly, run the [CMake configure command](/docs/user-guide/build/configure-and-build/) from the CMake CLI or CMake GUI. This registers everything you just added and emits errors if anything was misconfigured.
```

#### Step 2: Write new Python tests

> These tests primarily use standard [Python](https://docs.python.org/3/) code along with [PyTest](https://docs.pytest.org/). Depending on where your test is focused, it should likely include one of two automation libraries:

Reviewer: - Omit "primarily", unless there are other cases i which you use another Python version. If so, then consider mentioning them too. 
```suggestion
Python tests use standard [Python](https://docs.python.org/3/) code along with [PyTest](https://docs.pytest.org/). Depending on where your test is focused, it may include one of two automation libraries:
```

> Internal tests of O3DE Editor functionality should use [EditorPythonBindings](/docs/user-guide/testing/parallel-pattern/). This helps expose editor C++ interfaces to Python. These tests almost always require the `TEST_SERIAL` flag.
> 
> External tests at the Operating System level often use the [LyTestTools](/docs/user-guide/testing/lytesttools/) module, which simplifies interacting with O3DE applications. These tests often use the `TEST_SERIAL` flag.

Reviewer: - Consider putting into bullets and restructuring the sentences to name the automation libraries first. This is because the previous sentence introducing them says "it may include one of two automation libraries", so I was expecting to see the names of the libraries first. 
- I vaguely understand the difference between "almost always require" and "often use" regarding the `TEST_SERIAL` flag. I definitely think more information is needed to understand the context here, like: What does the `TEST_SERIAL` flag do? When do internal tests use them? When do external tests use them? You don't have to answer that here, maybe you go into it in another doc. But still I don't think it gives the reader much without the additional details. Maybe all they need to know in this doc is that "both internal and external tests may use the `TEST_SERIAL` flag. Specifying how often but leaving out a full explanation can cause confusion. 
```suggestion
- [EditorPythonBindings](/docs/user-guide/testing/parallel-pattern/) for internal tests on the O3DE Editor functionality. 

- [LyTestTools](/docs/user-guide/testing/lytesttools/) module for external tests at the Operating System level.

Both internal and external tests may use the `TEST_SERIAL` flag. 
```

title: Getting Started with Test Automation
description: Learn about the automated testing tools provided with Open 3D Engine (O3DE).
toc: true
weight: 500
---

This guide introduces the automated testing tools in **Open 3D Engine (O3DE)** and covers their basic usage. More detailed information on each tool exists on their specific documentation pages, linked in each section below.

## Overview

O3DE uses [CMake](https://cmake.org/cmake/help/latest/), a build system that includes [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html). CTest is a generic test runner tool which coordinates executing and reporting on tests, used by O3DE to start all automated testing. O3DE projects can also register their own tests through CTest, to test their unique game code or engine extensions. By running CTest locally, developers can verify code health before committing changes.

{{< important >}}
All CTest tests are executed in the O3DE Automated Review (AR) pipeline to help prevent bad merges. New code submissions are required to have tests; Changes to existing code must not cause regressions. Make sure to run your tests locally before any code submission to the O3DE project.

> {{< important >}}

Reviewer: Close the callout box
```suggestion
{{< /important >}}
```

O3DE provides wrapper code to help register tests for frameworks such as GoogleTest and PyTest into CTest. Along with the provided test-tools, these wrappers enable writing tests which can run on any operating system. XML files are generated to track the low-level results not reported to CTest, and artifacts are saved such as log output and crash dumps.

![CTest calls other test runners](/images/user-guide/testing/getting-started/ctest_to_runners.png)

In short, all tests in O3DE use CTest as a high-level test coordinator. When executed, CTest calls other lower-level test tools and then reports on their success.

### Starting CTest

CTest expects its working directory to be a CMake build directory, so be sure to first navigate a terminal to this directory. Then use `-C` to select a build configuration one you have already built locally:

> * **Windows**:
>     ```shell
>     cd <local_path_to>\\o3de\\build\\<build_folder>
>     ctest -C <build_configuration>
>     ```

Reviewer: ```suggestion
* **Windows**:
    ```shell
    cd <local_path_to>\o3de\build\<build_folder>
    ctest -C <build_configuration>
    ```
```
Wondering if it has to be double slash here? Other pages with Windows cd commands contain single slashes. E.g. [Create a new O3DE project](https://o3de.org/docs/welcome-guide/create/creating-projects-using-cli/creating-windows/#create-a-new-o3de-project)

* **Linux**:
    ```shell
    cd <local_path_to>/o3de/build/<build_folder>
    ctest -C <build_configuration>

>     ```

Reviewer: Originally there were some slight differences between Windows and Linux. But currently I don't see a difference except where the build folder is located and the back-/forward-slashes. Are there still some differences to highlight here or in the second shell sample? If not, we can simplify this by just listing the directory for Windows and Linux in the first step, and not the `ctest` commands in the first or second shell sample.

Author: I have no preference here, and made this change to fulfill the request of previous feedback. The point you bring up is why I had it as single examples, only becoming more explicit in one example.

{{< caution >}}
Without a filter the command will run every registered test, and likely result in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing `CTRL+C`.
{{< /caution >}}

CTest can also run a subset of labeled test suites with the `-L` argument. These test suites should be contained in parenthesis (`(..)`) and have their names separated with a `|` character. The following examples demonstrate this syntax for running the Main and Smoke suites for a `profile` build on Windows or Linux:

* **Windows**:
    ```shell
    cd C:\\o3de\\build\\vs_2019
    ctest -C profile -L "(SUITE_smoke|SUITE_main)"
    ```
* **Linux**:
    ```shell
    cd user/github/o3de/build/linux
    ctest -C profile -L "(SUITE_smoke|SUITE_main)"
    ```

> The Smoke and Main suites execute against pull requests in the AutomatedReview pipeline, and both are recommended for local verification. These suites are required to execute relatively fast, and must not intermittently fail.

Reviewer: ```suggestion
It's recommended that you verify your tests with the Main and Smoke suite on your local machine, as these tests will be executed on your pull request (PR), as part of the Automated Review pipeline in the `o3de` repository's PR workflow. For a successful PR, these suites must execute relatively fast and must not intermittently fail on your code. 
```
Rephrased some of this paragraph to clarify what the user is recommended to do (in "active" voice) and why.

Author: Some of the direction of implication has been changed. I'll rewrite to make this clearer.

Author: ```suggestion
It's recommended that you verify your tests with the Main and Smoke suite on your local machine. These tests will be executed on your pull request (PR), as part of the Automated Review pipeline in the `o3de` repository's PR workflow. Both of these suites must execute relatively fast and must not intermittently fail, and are an easy way to prove your change did not break other features. 
```

Reviewer: Great! Commit it

After running CTest, results save to `.../<build_folder>/Testing/`.  If you prefer to see full output of failures directly in your terminal, add the flag `--output-on-failure`.

For more information on CTest usage, refer to its [online documentation](https://cmake.org/cmake/help/latest/manual/ctest.1.html).

### Adding test modules to CTest

CTest registers entire modules of test code, which typically contain multiple individual tests of the same feature. To add new tests, complete a prerequisite and three major steps:

* **Prerequisite**: Add a build target for the production code which tests will target.

1. Add a build target for tests which need to be compiled (not required for Python).
2. Register your test module in a `CMakeLists.txt`.

> 3. Add individual test functions into the test module

Author: This was formatted due to the markdown linter suggesting that separate lists be separated. Joining the lists is a reasonable fix to your previous suggestion, will add this.

    FILES_CMAKE
        mymodule_files.cmake
    INCLUDE_DIRECTORIES
        PRIVATE
            Source
        PUBLIC
            Include
    BUILD_DEPENDENCIES
        PRIVATE
            Legacy::CryCommon
    COMPONENT
        MyModuleComponent
)

# New test module registration from Step 1
# this configures CMake to also build a library called MyModuleName.Tests.so (.dll on Windows)
# the rest of CMake will refer to this freshly created module by its NAME ("MyModuleName.Tests")
if(PAL_TRAIT_BUILD_SUPPORTS_TESTS)  # only create these test modules if the target platform supports testing

    ly_add_target(
        NAME MyModuleName.Tests SHARED
        NAMESPACE Gem
        FILES_CMAKE
            mymodule_test_files.cmake
        INCLUDE_DIRECTORIES
            PRIVATE
                Tests
                Source
        BUILD_DEPENDENCIES
            PRIVATE
                AZ::AzTest
                Legacy::CryCommon
                Gem::MyModuleName
        COMPONENT
            MyModuleComponent
    )

    # Register new test module from Step 2
    ly_add_googletest(
        NAME MyModuleName.Tests
    )

endif()
```

#### Prerequisite: Add a production build target

> Before configuring tests, you must first define the library that you want to test. If your tests target an existing feature, then this step is already completed and you simply need to find the correct file path. Target libraries are defined in a `CMakeLists.txt`, which typically lives in the same directory as the code or in a parent directory. For more information on configuring `CMakeLists.txt` files, please refer to [Getting Started with the CMake Build System](/docs/user-guide/build/). Note that you only need to define the production library that the tests target, not complete every feature in the production library. You can start configuring and writing tests before you finish working on the production code.

Author: The title of the index page for that section was used for the link title. I'm fine accepting this change, as it will be more resilient to changes to the linked page.

#### Step 1: Add a test build target

Similar to the production build target, the test target defines a library in a `CMakeLists.txt` configuration file. Start by finding the `CMakeLists.txt` that you created in the prerequisite step.  It should exist at a path similar to `o3de/.../<MyModule>/CMakeLists.txt`.

> Modify the `CMakeLists.txt` file to define your new test module with `ly_add_target()`. Similar to the production build target, it's easiest to create another file that lists the C++ files used to compile the test library. The example above uses `o3de/.../<MyModule>/mymodule_test_files.cmake`, which has content similar to the following:

Reviewer: Following up on this previous [comment](https://github.com/o3de/o3de.org/pull/1394/files/652f180c63330ac1259de5a222f35ca493002e53#r812424961), where we were looking to clarify that this instruction depends on the above **Prerequisite**, which links to the **[Build](https://deploy-preview-1394--o3deorg.netlify.app/docs/user-guide/build/)** section where it explains configuring `CMakeLists.txt` and creating `.cmake` files.

The confusion here is caused by a lack of info on `.cmake` files and their use in the Build section or anywhere in the O3DE documentation. This info is out-of-scope for this PR, but definitely needs to be addressed, so I'll file an issue for that. 

In the meantime, to clarify this step, simply stating ".cmake file" gives some hint and context. Also suggest putting last sentence in a new line to make it easier to scan descriptions of code samples. 
```suggestion
Modify the `CMakeLists.txt` file to define your new test module with `ly_add_target()`. Similar to the production build target, it's easiest to create another `.cmake` file that lists the C++ files used to compile the test library. 

The example above uses `o3de/.../<MyModule>/mymodule_test_files.cmake`, which has content similar to the following:
```

> Modify the `CMakeLists.txt` file to define your new test module with `ly_add_target()`. Similar to the production build target, it's easiest to create another file that lists the C++ files used to compile the test library. The example above uses `dev/.../<MyModule>/mymodule_test_files.cmake`, which has content similar to the following:

Author: It is a .cmake file, as explained in the build section. The Prerequisite step does not describe all steps or concepts from that section, and is intended to forward the user elsewhere.

> Modify the `CMakeLists.txt` file to define your new test module with `ly_add_target()`. Similar to the production build target, it's easiest to create another file that lists the C++ files used to compile the test library. The example above uses `dev/.../<MyModule>/mymodule_test_files.cmake`, which has content similar to the following:

Author: Proposed change was accepted, but un-resolving this to work on the clarification. Can this be made clearer about relying on the content learned in the prerequisite step?

```
set(FILES
    test/MyModuleTestFile.cpp
    test/MyModuleMathTests.cpp
)
```

#### Step 2: Register the test module

In `CMakeLists.txt`, register the module with CTest by using the helper function `ly_add_googletest()`.

> To verify everything is set up correctly, run the [CMake configure command](/docs/user-guide/build/configure-and-build/) from **CMake CLI** or **CMake GUI**. This registers everything you just added, and emits errors if anything is misconfigured.

Author: In the CMake GUI, there is a button labeled "Configure" and from the CLI you use a set of options which do not include flag `--build`

Where possible I was preferring to defer CMake knowledge to the CMake guide, and not dig too deeply into it here. Intent is to get the user looking at the actual build-system or CMake docs as quickly as possible.

> To verify everything is set up correctly, run the [CMake configure command](/docs/user-guide/build/configure-and-build/) from **CMake CLI** or **CMake GUI**. This registers everything you just added, and emits errors if anything is misconfigured.

Reviewer: Okay that sounds fine! Then I suggest at least making the link text more accurate to reflect what the user will actually find when they click on it. (I was expecting to see the actual CMake configure command immediately.)

```suggestion
To verify everything is set up correctly, run the CMake configure command from **CMake CLI** or **CMake GUI** (refer to [Configure and Build](/docs/user-guide/build/configure-and-build/) section). This registers everything you just added, and emits errors if anything is misconfigured.
```

While interpreted languages like Python can have low performance, tests should not perform heavy computation in Python code. Instead, tests should coordinate workflows by signalling events and then verifying a response. Leave heavy operations for the code targeted by the test, and perform simple checks in the Python test-code.

### Multiple Python Instances

Tests targeting O3DE can create multiple separate Python interpreter instances during execution, which each load different scripts. Most commonly, there are two instances: the _external interpreter_ and the _Editor interpreter_. The remainder of this section will help determine which environment your test should run in, and how to execute it there.

#### Tests in the external interpreter

> The external interpreter runs outside of any active O3DE application. This is the same interpreter that launches when you run `o3de/python/python.sh` or `o3de\python\python.cmd`, and is best used for tests which involve the following:

Author: Added, but retaining the backslash for the `.cmd` file since it is a windows-specific file. The separator was intended as a minor hint at this.

* Generic tests that launch applications and send external signals, often with [LyTestTools](/docs/user-guide/testing/lytesttools/).
* Monitoring application crashes.

To target PyTest in the external interpreter, by default your tests filenames should start with `test_`.

#### Tests in the Editor interpreter

The O3DE Editor internally manages a Python Interpreter, and exposes Editor-specific functionality through a Python bindings library. Although this environment isn't equivalent to launching `o3de/python/python.*`, it uses the same version of the Python interpreter. The Editor interpreter is best used for tests which involve the following:

* Target specific functionality within the Editor, using [EditorPythonBindings and EditorTest](/docs/user-guide/testing/parallel-pattern/).

> * Rely on external crash-handling (editor crash causes a test script crash).

Author: Rephrased and incorporated

To integrate with the Editor interpreter, create a test which uses `EditorPythonBindings`. These tests **must not** be in a file starting with `test_` or `tests_`, to avoid accidentally registering as failing tests.

{{< note >}}
PyTest is not used within the Editor interpreter, and as a consequence PyTest functionality is unavailable to tests which run in the Editor interpreter. Avoid dependencies on PyTest fixtures when designing these tests.

> EditorTest still uses PyTest to manage tests, and additionally handles external crash monitoring, batching, and parallelism. If this tool does not meet your needs, please reach out with a [feature request](https://github.com/o3de/o3de.org/issues/new/choose) or start a discussion in [Discord](https://discord.gg/p3padwr58u) channel SIG-Testing!

Author: The latter edit sounds artificial to me. Why does this need to spell out Special Interest Group?

> EditorTest still uses PyTest to manage tests, and additionally handles external crash monitoring, batching, and parallelism. If this tool does not meet your needs, please reach out with a [feature request](https://github.com/o3de/o3de.org/issues/new/choose) or start a discussion in [Discord](https://discord.gg/p3padwr58u) channel SIG-Testing!

Reviewer: Yeah I see it feels like we're cramming in the name at the end there. The reason is "SIG" is an abbreviation, so we have to clarify them when they first appear in the document. It's also the official name of the SIG, so gives readers an introduction to them. 
Another way you can say it is: 
>If this tool does not meet your needs, please reach out with a [feature request](https://github.com/o3de/o3de.org/issues/new/choose) or start a discussion with the Testing Special Interest Group in the [Discord](https://discord.gg/p3padwr58u) channel, sig-testing!

I kept "sig-testing" lowercase as that's how it appears in Discord.

{{< /note >}}

### Registering a new Python test

> Registering a Python-based test is simpler than registering a C++ test. However, it still requires that you define the C++ library that you want to test, which is likely already completed before designing integration tests. The steps below assume the production library is already defined. You can read more about defining production code in [Getting Started with the CMake Build System](/docs/user-guide/build/).

Reviewer: ```suggestion
Registering a Python-based test is simpler than registering a C++ test. However, it still requires that you define the C++ library that you want to test, which you've likely already completed before designing integration tests. The steps below assume the production library is already defined. You can read more about defining production code in the [Build](/docs/user-guide/build/) section.
```

Similar to previous suggestion, is there a more specific section you're referring to? If so, update the link. You can also specify subsections within a page by adding `#` and the name of the subsection, lower-cased and space-separated with hyphens. For example: `/docs/user-guide/build/configure-and-build#configuring-projects`

For now, my suggestion updates the name of the reference link to more accurately describe what users will find when they click on it.

Author: I do not want to deep-link into this other section of documentation, and users of this system need to know the content it introduces.

#### Step 1: Register a PyTest target

Find the `CMakeLists.txt` that defines the system you are testing. The file should exist at a path similar to `o3de/.../<MyModule>/CMakeLists.txt`. Depending on the test, it may be more appropriate to register the test to a `CMakeLists.txt` file at another location. Reasons include:

* A test for a sub-module should exist at a child directory of its module.
* A broad test integrating across multiple features should exist in a parent directory of the features.

> * A test which relies on code or assets in a game project must exist in that game project.

Reviewer: ```suggestion
* A test that relies on code or assets in a game project must exist in that game project.
```

  * This avoids the case where disabling the project breaks the still-registered tests.

When you find the right `CMakeLists.txt` to register the test, add a line similar to the following:

```
ly_add_pytest(
    NAME myPythonTest 
    PATH ${CMAKE_CURRENT_LIST_DIR}/Tests/test_MyFile.py
    TEST_SERIAL # many larger-than-unit tests can interfere with one another in parallel
    RUNTIME_DEPENDENCIES # helps the test build-and-run from an IDE
        Legacy::Editor
        AZ::AssetProcessor
        AutomatedTesting.Assets
)
```

> To verify everything is set up correctly, run the [CMake configure command](/docs/user-guide/build/configure-and-build/) from the CMake CLI or CMake GUI. This registers everything you just added and emits errors if anything was misconfigured.

Reviewer: Similar to previous comment, can you specify the link to "CMake configure command", as there are numerous command samples in the linked **Configure and Build** page? Also, if its a single command, consider writing the command here, so users don't need to redirect to another page to find the command. You can keep the link to the "CMake configure command" page so users can find the thorough instructions there, and for the steps on how to do it via CMake GUI.

Author: Similarly, I do not want to describe the content from another section.

Reviewer: Similar to response above, I suggest at least making the link text more accurate to reflect what the user will actually find when they click on it.

```suggestion
To verify everything is set up correctly, run the CMake configure command from **CMake CLI** or **CMake GUI** (refer to the [Configure and Build](/docs/user-guide/build/configure-and-build/) section). This registers everything you just added, and emits errors if anything is misconfigured.
```

#### Step 2: Write new Python tests

> These tests use [Python 3](https://docs.python.org/3/) code along with [PyTest](https://docs.pytest.org/). Depending on the feature your test is focuses on, the test may include one of two automation libraries provided with O3DE:

Reviewer: ```suggestion
Python-based tests use [Python 3](https://docs.python.org/3/) code along with [PyTest](https://docs.pytest.org/). Depending on the feature your test is focused on, the test may include one of two automation libraries provided with O3DE:
```

* [EditorPythonBindings](/docs/user-guide/testing/parallel-pattern/) for internal tests of the O3DE Editor functionality.
  * These tests always require the `TEST_SERIAL` CTest flag.

> * [LyTestTools](/docs/user-guide/testing/lytesttools/) module for external tests at the Operating System level.

Reviewer: ```suggestion
* [LyTestTools](/docs/user-guide/testing/lytesttools/) module for external tests at the operating system level.
```

O3DE provides wrapper code to help register tests for frameworks such as GoogleTest and PyTest into CTest. Along with the provided test-tools, these wrappers enable writing tests that can run on any operating system supported by the underlying frameworks. XML files are generated to track the low-level results not reported to CTest, and artifacts are saved such as log output and crash dumps.

![CTest calls other test runners](/images/user-guide/testing/getting-started/ctest_to_runners.png)

In short, all tests in O3DE use CTest as a high-level test coordinator. When executed, CTest calls other lower-level test tools and then reports on their success.

### Starting CTest

CTest expects its working directory to be a CMake build directory, so be sure to first navigate to this directory in a terminal. Then, use `-C` to select a build configuration, one that you've already built locally:

> * **Windows**:
>     ```shell
>     cd <local_path_to>\o3de\build\<build_folder>
>     ctest -C <build_configuration>
>     ```
> 
> * **Linux**:
>     ```shell
>     cd <local_path_to>/o3de/build/<build_folder>
>     ctest -C <build_configuration>
>     ```

Reviewer: ```suggestion
* **Windows**:
    ```cmd
    cd <local_path_to>\o3de\build\<build_folder>
    ctest -C <build_configuration>
    ```

* **Linux**:
    ```cmd
    cd <local_path_to>/o3de/build/<build_folder>
    ctest -C <build_configuration>
    ```
```

{{< caution >}}
Without a filter the command will run every registered test, and likely result in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing `CTRL+C`.
{{< /caution >}}

CTest can also run a subset of labeled test suites with the `-L` argument. These test suites should be contained in parenthesis (`(..)`) and have their names separated with a `|` character. The following examples demonstrate this syntax for running the Main and Smoke suites for a `profile` build on Windows or Linux:

* **Windows**:

>     ```shell
>     cd C:\o3de\build\vs_2019
>     ctest -C profile -L "(SUITE_smoke|SUITE_main)"
>     ```
> * **Linux**:
>     ```shell
>     cd user/github/o3de/build/linux

Reviewer: ```suggestion
    ```cmd
    cd C:\o3de\build\vs_2019
    ctest -C profile -L "(SUITE_smoke|SUITE_main)"
    ```
* **Linux**:
    ```cmd
    cd user/github/o3de/build/linux
```

O3DE provides wrapper code to help register tests for frameworks such as GoogleTest and PyTest into CTest. Along with the provided test-tools, these wrappers enable writing tests which can run on any operating system. XML files are generated to track the low-level results not reported to CTest, and artifacts are saved such as log output and crash dumps.

![CTest calls other test runners](/images/user-guide/testing/getting-started/ctest_to_runners.png)

In short, all tests in O3DE use CTest as a high-level test coordinator. When executed, CTest calls other lower-level test tools and then reports on their success.

### Starting CTest

CTest expects its working directory to be a CMake build directory, so be sure to first navigate a terminal to this directory. Then use `-C` to select a build configuration one you have already built locally:

* **Windows**:
    ```shell
    cd <local_path_to>\\o3de\\build\\<build_folder>
    ctest -C <build_configuration>
    ```

* **Linux**:
    ```shell
    cd <local_path_to>/o3de/build/<build_folder>
    ctest -C <build_configuration>

>     ```

Author: I agree that the change of directory can be omitted after the first example. I still want to provide at least one concrete path in an example, unless there is a reason not to.

{{< caution >}}
Without a filter the command will run every registered test, and likely result in multiple hours of test execution! If you ever want to stop CTest, send an interrupt signal by selecting the terminal and pressing `CTRL+C`.
{{< /caution >}}

CTest can also run a subset of labeled test suites with the `-L` argument. These test suites should be contained in parenthesis (`(..)`) and have their names separated with a `|` character. The following examples demonstrate this syntax for running the Main and Smoke suites for a `profile` build on Windows or Linux:

* **Windows**:
    ```shell

>     cd C:\\o3de\\build\\vs_2019

Author: I intentionally wanted to provide at least one example that did not use relative path placeholders.

>     cd C:\\o3de\\build\\vs_2019

Author: Committing the following change, so that the relative paths can be discussed in isolation:

```suggestion
    cd C:\\o3de\\build\\vs_2019
```

    ctest -C profile -L "(SUITE_smoke|SUITE_main)"
    ```
* **Linux**:
    ```shell

>     cd user/github/o3de/build/linux

Author: Ditto, I think one of these should remain a concrete example path. Does this need a better supporting comment?
